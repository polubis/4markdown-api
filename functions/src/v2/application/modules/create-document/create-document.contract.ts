import { type PrivateDocumentModel } from '@domain/models/document';
import {
  documentCodeSchema,
  documentNameSchema,
} from '@utils/document-schemas';
import { type Id } from '@utils/validators';
import { z } from 'zod';

const createDocumentPayloadSchema = z.object({
  name: documentNameSchema,
  code: documentCodeSchema,
});

type CreateDocumentPayload = z.infer<typeof createDocumentPayloadSchema>;
type CreateDocumentDto = PrivateDocumentModel & { id: Id; authorId: Id };

export { createDocumentPayloadSchema };
export type { CreateDocumentDto, CreateDocumentPayload };
