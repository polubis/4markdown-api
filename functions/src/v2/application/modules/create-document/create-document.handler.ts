import { nowISO, uuid } from '@libs/helpers/stamps';
import type { DocumentModel, DocumentsModel } from '@domain/models/document';
import { errors } from '@utils/errors';
import type {
  CreateDocumentDto,
  CreateDocumentPayload,
} from './create-document.contract';
import { type ProtectedControllerHandlerContext } from '@utils/controller';

const createDocumentHandler = async ({
  payload,
  context: { db, uid },
}: {
  payload: CreateDocumentPayload;
  context: ProtectedControllerHandlerContext;
}) => {
  const documentsRef = db.collection(`docs`).doc(uid);
  const documentsSnap = await documentsRef.get();

  const documents = documentsSnap.data() as DocumentsModel | undefined;

  const id = uuid();
  const cdate = nowISO();

  const documentModel: DocumentModel = {
    cdate,
    mdate: cdate,
    name: payload.name.raw,
    code: payload.code,
    path: payload.name.path,
    visibility: `private`,
  };

  const documentsModel: DocumentsModel = {
    [id]: documentModel,
  };

  const dto: CreateDocumentDto = {
    ...documentModel,
    id,
  };

  if (!documents) {
    await documentsRef.set(documentsModel);
    return dto;
  }

  const exists = Object.values(documents).some(
    (document) => document.path === payload.name.path,
  );

  if (exists) {
    throw errors.exists(`Document with ${payload.name.raw} already exists`);
  }

  await documentsRef.update(documentsModel);

  return dto;
};

export { createDocumentHandler };
