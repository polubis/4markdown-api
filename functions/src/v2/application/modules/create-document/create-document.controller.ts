import { protectedController } from '@utils/controller';
import { parse } from '@utils/parse';
import {
  createDocumentPayloadSchema,
  type CreateDocumentDto,
} from './create-document.contract';
import { createDocumentHandler } from './create-document.handler';

const createDocumentController = protectedController<CreateDocumentDto>(
  async (rawPayload, context) => {
    return await createDocumentHandler({
      context,
      payload: await parse(createDocumentPayloadSchema, rawPayload),
    });
  },
);

export { createDocumentController };
