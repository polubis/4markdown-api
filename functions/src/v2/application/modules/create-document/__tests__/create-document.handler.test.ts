import { createDocumentHandler } from '../create-document.handler';
import {
  CreateDocumentDto,
  CreateDocumentPayload,
} from '../create-document.contract';
import {
  DocumentModelVisibility,
  DocumentsModel,
} from '@domain/models/document';

jest.mock(`@libs/helpers/stamps`, () => ({
  nowISO: jest.fn(() => `2024-01-01T00:00:00Z`),
  uuid: jest.fn(() => `mock-id`),
}));

describe(`Document creation works when`, () => {
  const validPayload: CreateDocumentPayload = {
    name: {
      raw: `Test Document For Me`,
      slug: `test-document-for-me`,
      path: `/test-document-for-me/`,
      segments: [`test`, `document`, `for`, `me`],
    },
    code: `console.log("test")`,
  };
  const projectId = `project-id`;
  const uid = `user-uid`;

  const validDto: CreateDocumentDto = {
    id: `mock-id`,
    authorId: uid,
    cdate: `2024-01-01T00:00:00Z`,
    mdate: `2024-01-01T00:00:00Z`,
    name: validPayload.name.raw,
    code: validPayload.code,
    path: validPayload.name.path,
    visibility: DocumentModelVisibility.Private,
  };

  it(`creates first document via "set" function from db provider and ignores "update"`, async () => {
    const setSpy = jest.fn();
    const updateSpy = jest.fn();

    const result = await createDocumentHandler({
      payload: validPayload,
      context: {
        uid,
        db: {
          collection: () => ({
            doc: () => ({
              get: async () => ({
                data: () => undefined,
              }),
              set: setSpy,
              update: updateSpy,
            }),
          }),
        } as any,
        projectId,
      },
    });

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { id, authorId, ...documentModel } = validDto;
    const documentsModel: DocumentsModel = {
      [id]: documentModel,
    };

    expect(result).toEqual(validDto);
    expect(setSpy).toHaveBeenCalledWith(documentsModel);
    expect(updateSpy).not.toHaveBeenCalled();
  });

  it(`throws an error when document with provided name exist`, async () => {
    const { id, ...documentModel } = validDto;
    const documentsModel: DocumentsModel = {
      [id]: documentModel,
    };

    try {
      await createDocumentHandler({
        payload: validPayload,
        context: {
          uid,
          db: {
            collection: () => ({
              doc: () => ({
                get: async () => ({
                  data: () => documentsModel,
                }),
              }),
            }),
          } as any,
          projectId,
        },
      });
    } catch (error: unknown) {
      expect(error).toMatchSnapshot();
    }
  });

  it(`adds new document to existing documents collection`, async () => {
    const setSpy = jest.fn();
    const updateSpy = jest.fn();

    const result = await createDocumentHandler({
      payload: validPayload,
      context: {
        uid,
        db: {
          collection: () => ({
            doc: () => ({
              get: async () => ({
                data: () => ({
                  'random-id': {
                    id: `mock-id-2`,
                    cdate: `2024-01-01T00:00:00Z`,
                    mdate: `2024-01-01T00:00:00Z`,
                    name: `My custom document name`,
                    code: `empty code`,
                    path: `/my-custom-document-name/`,
                    visibility: `private`,
                  },
                }),
              }),
              set: setSpy,
              update: updateSpy,
            }),
          }),
        } as any,
        projectId,
      },
    });

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { id, authorId, ...documentModel } = validDto;
    const documentsModel: DocumentsModel = {
      [id]: documentModel,
    };

    expect(result).toEqual(validDto);
    expect(setSpy).not.toHaveBeenCalled();
    expect(updateSpy).toHaveBeenCalledWith(documentsModel);
  });
});
