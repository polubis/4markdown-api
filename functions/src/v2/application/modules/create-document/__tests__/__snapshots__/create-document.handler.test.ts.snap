// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document creation works when throws an error when document with provided name exist 1`] = `[Error: {"symbol":"already-exists","content":"Document with Test Document For Me already exists","message":"Document with Test Document For Me already exists"}]`;
